File: 543-DiameterOfBinaryTree/543-DiameterOfBinaryTree.java
// Last updated: 8/16/2026, 9:50:14 PM
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxDepth = 0;
    public int diameterOfBinaryTree(TreeNode root) {
        // for every node, the 'diameter' which "PASS/CROSS it" = left.depth + right.depth
        // use 'maxDepth' to record current max, cause 'diameter' may not cross root
        if (root == null) return 0;

        getTreeDepth(root);
        return maxDepth;
    }

    private int getTreeDepth(TreeNode root) {
        // calculdate the depth of a treenode root
        // and update maxDepth value
        if (root == null) return 0;
        int maxLeft = getTreeDepth(root.left);
        int maxRight = getTreeDepth(root.right);

        maxDepth = Math.max(maxLeft + maxRight, maxDepth);

        return Math.max(maxLeft, maxRight) + 1;
    }
}

File: 543-DiameterofBinaryTree/543-DiameterofBinaryTree.py
# Last updated: 8/17/2026, 5:06:51 PM
1# Definition for a binary tree node.
2# class TreeNode:
3#     def __init__(self, val=0, left=None, right=None):
4#         self.val = val
5#         self.left = left
6#         self.right = right
7class Solution:
8    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
9        # res = max(left_depth + right_depth)
10        # dfs to get tre depth, and update global variable 'res'
11
12        max_diameter = 0
13
14        def height(node: Optional[TreeNode]) -> int:
15            nonlocal max_diameter # ensure max_diameter is read from global
16            
17            if not node:
18                return 0
19            
20            l_h = height(node.left)
21            r_h = height(node.right)
22
23            max_diameter = max(max_diameter, l_h + r_h)
24
25            return max(l_h, r_h) + 1
26        
27        height(root)
28
29        return max_diameter